File: Raspberry Pi Embedded - Python/dynamixel.py
from dynamixel_sdk import *  # Import the Dynamixel SDK library
import time

class DxlServo:
    ADDR_POSITION_P_GAIN = 84  # Address for P gain
    ADDR_POSITION_I_GAIN = 82  # Address for I gain
    ADDR_POSITION_D_GAIN = 80  # Address for D gain
    ADDR_PROFILE_VELOCITY = 112  # Address for Profile Velocity
    ADDR_PROFILE_ACCELERATION = 108  # Address for Profile Acceleration
    ADDR_GOAL_VELOCITY = 104  # Address for velocity control
    ADDR_OPERATING_MODE = 11  # Address for setting Operating Mode
    ADDR_GOAL_POSITION = 116  # Control table address for goal position
    ADDR_PRESENT_POSITION = 132  # Control table address for present position
    ADDR_TORQUE_ENABLE = 64  # Address for enabling torque
    OPERATING_MODE_TORQUE = 0     # Current Control Mode
    OPERATING_MODE_VELOCITY = 1
    OPERATING_MODE_POSITION = 3
    TORQUE_ENABLE = 1
    TORQUE_DISABLE = 0
    ADDR_GOAL_PWM = 100
    ADDR_PRESENT_PWM = 124

    def __init__(self, dxl_id, offset, speeds, multiplier=1, operating_mode=1):
        self.PORT_NAME = '/dev/ttyUSB0'
        self.BAUD_RATE = 57600
        self.PROTOCOL_VERSION = 2.0
        self.DXL_ID = dxl_id
        self.operating_mode = operating_mode
        self.offset = offset
        self.multiplier = multiplier
        self.MIN_POSITION = 0  # Min position value (0 degrees)
        self.MAX_POSITION = 4095  # Max position value (360 degrees)
        self.offset_center = offset * 4095 / 360.0
        #print("Offset center is: " + str(self.offset_center))
        self.speeds = speeds
        self.last_set_torque = None

        # Create Port and Packet Handlers
        self.port_handler = PortHandler(self.PORT_NAME)
        self.packet_handler = PacketHandler(self.PROTOCOL_VERSION)

        # Initialize the port
        self.initialize()
        self.disable_torque()
        self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_OPERATING_MODE, self.operating_mode)
        self.enable_torque()
        #self.set_velocity_and_acceleration(50, 10)

    def initialize(self):
        """Initializes the communication port."""
        if not self.port_handler.openPort():
            raise Exception("Failed to open the port.")
        if not self.port_handler.setBaudRate(self.BAUD_RATE):
            raise Exception("Failed to set baudrate.")
        #print("Port initialized successfully.")

    def torque_mode(self):
        self.disable_torque()
        dxl_comm_result, dxl_error = self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_OPERATING_MODE, 16)
        if dxl_comm_result != COMM_SUCCESS:
            print(f"Failed to set Current Control Mode: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
            return False
        elif dxl_error:
            print(f"Error setting mode: {self.packet_handler.getRxPacketError(dxl_error)}")
        self.enable_torque()

    def set_torque(self, torque):
        if torque == self.last_set_torque:
            return
        torque_constant = 1.78 # Nm/A
        current = torque / torque_constant
        dyna_units = int(current / .00269) 
        print(f"Set torque {torque:.2f} ({dyna_units} in dxl units)")
        self.last_set_torque = torque
        dxl_comm_result, dxl_error = self.packet_handler.write2ByteTxRx(self.port_handler, self.DXL_ID, 102, dyna_units)
        if dxl_comm_result != COMM_SUCCESS:
            print(f"Failed to set goal PWM: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
        elif dxl_error:
            print(f"Error setting goal PWM: {self.packet_handler.getRxPacketError(dxl_error)}")

    def set_pid_gains(self, p_gain, i_gain, d_gain):
        """Sets the PID gains for the servo."""
        # Set P Gain
        dxl_comm_result, dxl_error = self.packet_handler.write2ByteTxRx(
            self.port_handler, self.DXL_ID, self.ADDR_POSITION_P_GAIN, p_gain
        )
        if dxl_comm_result != COMM_SUCCESS:
            raise Exception(f"P Gain Error: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
        if dxl_error != 0:
            raise Exception(f"P Gain Error: {self.packet_handler.getRxPacketError(dxl_error)}")
        print(f"P Gain set to {p_gain}.")

        # Set I Gain
        dxl_comm_result, dxl_error = self.packet_handler.write2ByteTxRx(
            self.port_handler, self.DXL_ID, self.ADDR_POSITION_I_GAIN, i_gain
        )
        if dxl_comm_result != COMM_SUCCESS:
            raise Exception(f"I Gain Error: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
        if dxl_error != 0:
            raise Exception(f"I Gain Error: {self.packet_handler.getRxPacketError(dxl_error)}")
        print(f"I Gain set to {i_gain}.")

        # Set D Gain
        dxl_comm_result, dxl_error = self.packet_handler.write2ByteTxRx(
            self.port_handler, self.DXL_ID, self.ADDR_POSITION_D_GAIN, d_gain
        )
        if dxl_comm_result != COMM_SUCCESS:
            raise Exception(f"D Gain Error: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
        if dxl_error != 0:
            raise Exception(f"D Gain Error: {self.packet_handler.getRxPacketError(dxl_error)}")
        print(f"D Gain set to {d_gain}.")

    def enable_torque(self):
        """Enables torque on the servo."""
        self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_TORQUE_ENABLE, self.TORQUE_ENABLE)

    def disable_torque(self):
        """Disables torque on the servo."""
        self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_TORQUE_ENABLE, self.TORQUE_DISABLE)

    def get_present_position(self):
        """Reads and returns the present position of the servo."""
        present_position, _, _ = self.packet_handler.read4ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_PRESENT_POSITION)
        if present_position > 2**31:
            present_position = self.MAX_POSITION + present_position - 2**32
        #print("Curr position: " + str(present_position))
        if present_position > 4095:
            present_position = present_position - 4095
        return present_position
    
    def get_present_position_degrees(self):
        """Reads and returns the present position of the servo."""
        present_position = self.get_present_position()
        return self.convert_servo_position_to_degrees(present_position)

    def set_velocity_and_acceleration(self, velocity, acceleration):
        """Sets the profile velocity and acceleration for smooth movement."""
        self.packet_handler.write4ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_PROFILE_VELOCITY, velocity)
        self.packet_handler.write4ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_PROFILE_ACCELERATION, acceleration)
        print(f"Profile Velocity set to {velocity}, Acceleration set to {acceleration}.")
    
    def move_to_with_velocity(self, goal_position):
        """Moves the servo smoothly to the target using velocity control with velocity ramping."""
        stop_threshold = 10

        goal_position = self.convert_degrees_to_servo_position(goal_position)

        # Switch to velocity mode
        # self.disable_torque()
        # self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_OPERATING_MODE, self.OPERATING_MODE_VELOCITY)
        # self.enable_torque()

        print(f"Moving servo {self.DXL_ID} to {goal_position} with velocity control.")

        while True:
            present_position = self.get_present_position()
            raw_diff = goal_position - present_position

            # Wrap the difference so it falls within [-2048, 2047]
            if raw_diff > self.MAX_POSITION // 2:
                raw_diff -= self.MAX_POSITION
            elif raw_diff < -(self.MAX_POSITION // 2):
                raw_diff += self.MAX_POSITION

            error = abs(raw_diff)

            new_velocity = self.speeds[0]
            if error < 100:
                new_velocity = self.speeds[2]
            elif error < 250:
                new_velocity = self.speeds[1]

            if raw_diff < 0:
                new_velocity = new_velocity * -1

            # Stop if within final threshold
            if error < stop_threshold:
                break

            # Write velocity command
            self.packet_handler.write4ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_GOAL_VELOCITY, new_velocity)

            print(f"Moving... Current Position: {present_position}, Target: {goal_position}, Error: {error}, Velocity: {new_velocity}")
            time.sleep(0.1)

        # Stop movement immediately (no gradual deceleration)
        self.packet_handler.write4ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_GOAL_VELOCITY, 0)
        print(f"Now at {self.get_present_position()}, stopping.")

        # Switch back to position mode
        # self.disable_torque()
        # self.packet_handler.write1ByteTxRx(self.port_handler, self.DXL_ID, self.ADDR_OPERATING_MODE, self.OPERATING_MODE_POSITION)
        # self.enable_torque()

    def convert_servo_position_to_degrees(self, servo_position):
        """Converts the servo's internal position scale back to degrees."""
        #print("offset center: " + str(self.offset_center))
        servo_values_per_degree = self.MAX_POSITION / 360.0
        #print("multiplier: " + str(self.multiplier))
        #print("numer: " + str(servo_position - self.offset_center))
        #print("denom: " + str(servo_values_per_degree * self.multiplier))
        #difference = servo_position - self.offset_center

        result = (servo_position - self.offset_center) / (servo_values_per_degree * self.multiplier)
        if result > 180.0:
            result = result - 360.0
        return round(result, 2)
    
    def set_position(self, pos_deg, velocity=None):
        if self.operating_mode != self.OPERATING_MODE_POSITION:
            raise RuntimeError("Servo is not in position control mode.")
        
        if velocity is not None:
            velocity = int(velocity)
            self.packet_handler.write4ByteTxRx(
                self.port_handler, self.DXL_ID, self.ADDR_PROFILE_VELOCITY, velocity
            )

        self.packet_handler.write4ByteTxRx(
            self.port_handler, self.DXL_ID, self.ADDR_PROFILE_ACCELERATION, 50
        )

        goal_position = self.convert_degrees_to_servo_position(pos_deg)
        dxl_comm_result, dxl_error = self.packet_handler.write4ByteTxRx(
            self.port_handler, self.DXL_ID, self.ADDR_GOAL_POSITION, goal_position
        )

        if dxl_comm_result != COMM_SUCCESS:
            raise Exception(f"Failed to set goal position: {self.packet_handler.getTxRxResult(dxl_comm_result)}")
        if dxl_error != 0:
            raise Exception(f"Error from servo: {self.packet_handler.getRxPacketError(dxl_error)}")


    def close(self):
        """Closes the communication port."""
        self.port_handler.closePort()
        print("Port closed.")

    def convert_degrees_to_servo_position(self, pos):
        """Converts degrees to the servo's internal position scale."""
        servo_values_per_degree = self.MAX_POSITION / 360.0
        result = int(self.offset_center + (pos * servo_values_per_degree) * self.multiplier)
        if result < 0:
            result = result + self.MAX_POSITION
        return result

# Example Usage
if __name__ == "__main__":
    servo = DxlServo(1, 0)  # Initialize servo with ID 1

    print("Moving using position control:")
    servo.move_to_degrees(0)
    time.sleep(2)
    servo.move_to_degrees(-90)
    time.sleep(2)
    servo.move_to_degrees(90)

    print("Moving using velocity control:")
    servo.move_to_with_velocity(0, max_velocity=40, stop_threshold=10)
    time.sleep(2)
    servo.move_to_with_velocity(-90, max_velocity=40, stop_threshold=10)
    time.sleep(2)
    servo.move_to_with_velocity(90, max_velocity=40, stop_threshold=10)

    servo.close()

File: Raspberry-Pi-Embedded-C++/dynamixel.cpp
#include "dynamixel.hpp"
#include <dynamixel_sdk/dynamixel_sdk.h>
#include <string>
#include <array>
#include <iostream>
#include <thread>
#include <chrono>
#include <cmath>
#include <limits>

// Method implementations for DxlServo

// Constructor
DxlServo::DxlServo(
    uint8_t dxl_id,
    double offset_deg,
    std::array<int,3> speeds,
    double multiplier,
    uint8_t operating_mode,
    const std::string& port_name,
    int baud_rate,
    float protocol_version)
    : PORT_NAME(port_name),
      BAUD_RATE(baud_rate),
      PROTOCOL_VERSION(protocol_version),
      DXL_ID(dxl_id),
      operating_mode_(operating_mode),
      offset_(offset_deg),
      multiplier_(multiplier),
      MIN_POSITION(0),
      MAX_POSITION(4095),
      offset_center_(0.0),
      speeds_(speeds),
      last_set_torque_(std::numeric_limits<double>::quiet_NaN()),
      portHandler_(nullptr),
      packetHandler_(nullptr)
{
    offset_center_ = offset_ * 4095.0 / 360.0;

    // Initialize SDK handlers
    portHandler_   = dynamixel::PortHandler::getPortHandler(PORT_NAME.c_str());
    packetHandler_ = dynamixel::PacketHandler::getPacketHandler(PROTOCOL_VERSION);

    initialize();
    disable_torque();

    // Set initial operating mode
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler_->write1ByteTxRx(portHandler_, DXL_ID, ADDR_OPERATING_MODE, operating_mode_, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        throw std::runtime_error(std::string("Failed to set operating mode: ") + packetHandler_->getTxRxResult(dxl_comm_result));
    }
    if (dxl_error) {
        throw std::runtime_error(std::string("Servo error setting operating mode: ") + packetHandler_->getRxPacketError(dxl_error));
    }

    enable_torque();
}

// Destructor
DxlServo::~DxlServo() {
    try { close(); } catch (...) {}
}

void DxlServo::initialize() {
    if (!portHandler_->openPort())
        throw std::runtime_error("Failed to open port: " + PORT_NAME);
    if (!portHandler_->setBaudRate(BAUD_RATE))
        throw std::runtime_error("Failed to set baud rate: " + std::to_string(BAUD_RATE));
}

void DxlServo::torque_mode() {
    // Mirrors Python: writes 16 to Operating Mode (note: that’s actually PWM mode)
    disable_torque();
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler_->write1ByteTxRx(
        portHandler_, DXL_ID, ADDR_OPERATING_MODE, 16, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        std::cerr << "Failed to set Current Control Mode: "
                  << packetHandler_->getTxRxResult(dxl_comm_result) << "\n";
        return;
    } else if (dxl_error) {
        std::cerr << "Error setting mode: "
                  << packetHandler_->getRxPacketError(dxl_error) << "\n";
    }
    enable_torque();
}

void DxlServo::set_torque(double torque_nm) {
    if (!std::isnan(last_set_torque_) && torque_nm == last_set_torque_) return;

    // Python logic: torque -> current -> dxl units (0.00269 A per unit), write to address 102
    const double torque_constant = 1.78; // Nm/A
    double current_a = torque_nm / torque_constant;
    int dyna_units = static_cast<int>(current_a / 0.00269);

    std::cout << "Set torque " << torque_nm << " (" << dyna_units << " in dxl units)\n";
    last_set_torque_ = torque_nm;

    uint8_t dxl_error = 0;
    // Note: Python writes to address 102 specifically (not ADDR_GOAL_PWM=100)
    int dxl_comm_result = packetHandler_->write2ByteTxRx(
        portHandler_, DXL_ID, 102, dyna_units, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        std::cerr << "Failed to set goal PWM: "
                  << packetHandler_->getTxRxResult(dxl_comm_result) << "\n";
    } else if (dxl_error) {
        std::cerr << "Error setting goal PWM: "
                  << packetHandler_->getRxPacketError(dxl_error) << "\n";
    }
}

void DxlServo::set_pid_gains(uint16_t p_gain, uint16_t i_gain, uint16_t d_gain) {
    write2B(ADDR_POSITION_P_GAIN, p_gain, "P Gain");
    std::cout << "P Gain set to " << p_gain << ".\n";
    write2B(ADDR_POSITION_I_GAIN, i_gain, "I Gain");
    std::cout << "I Gain set to " << i_gain << ".\n";
    write2B(ADDR_POSITION_D_GAIN, d_gain, "D Gain");
    std::cout << "D Gain set to " << d_gain << ".\n";
}

void DxlServo::enable_torque() {
    uint8_t dxl_error = 0;
    packetHandler_->write1ByteTxRx(portHandler_, DXL_ID, ADDR_TORQUE_ENABLE, TORQUE_ENABLE, &dxl_error);
}

void DxlServo::disable_torque() {
    uint8_t dxl_error = 0;
    packetHandler_->write1ByteTxRx(portHandler_, DXL_ID, ADDR_TORQUE_ENABLE, TORQUE_DISABLE, &dxl_error);
}

int DxlServo::get_present_position() {
    uint8_t dxl_error = 0;
    uint32_t pos = 0;
    int dxl_comm_result = packetHandler_->read4ByteTxRx(
        portHandler_, DXL_ID, ADDR_PRESENT_POSITION, &pos, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        throw std::runtime_error(std::string("Read position failed: ")
                                 + packetHandler_->getTxRxResult(dxl_comm_result));
    }
    if (dxl_error) {
        throw std::runtime_error(std::string("Servo error on read: ")
                                 + packetHandler_->getRxPacketError(dxl_error));
    }

    // Mirror Python’s wrap logic
    int present_position = static_cast<int>(pos);
        if (present_position > (1 << 31)) {
            present_position = MAX_POSITION + present_position - (1ULL << 32);
    }
    if (present_position > 4095) {
        present_position -= 4095;
    }
    return present_position;
}

double DxlServo::get_present_position_degrees() {
    return convert_servo_position_to_degrees(get_present_position());
}

void DxlServo::set_velocity_and_acceleration(uint32_t velocity, uint32_t acceleration) {
    write4B(ADDR_PROFILE_VELOCITY,     velocity,     "Profile Velocity");
    write4B(ADDR_PROFILE_ACCELERATION, acceleration, "Profile Acceleration");
    std::cout << "Profile Velocity set to " << velocity
              << ", Acceleration set to " << acceleration << ".\n";
}

void DxlServo::move_to_with_velocity(double goal_deg) {
    const int stop_threshold = 10; // same as Python
    int goal_position = convert_degrees_to_servo_position(goal_deg);

    std::cout << "Moving servo " << int(DXL_ID) << " to " << goal_position
              << " with velocity control.\n";

    while (true) {
        int present_position = get_present_position();
        int raw_diff = goal_position - present_position;

        // Wrap to [-2048, 2047]
        if (raw_diff > (MAX_POSITION / 2)) raw_diff -= MAX_POSITION;
        else if (raw_diff < -(MAX_POSITION / 2)) raw_diff += MAX_POSITION;

        int error = std::abs(raw_diff);

        int new_velocity = speeds_[0];
        if (error < 100)      new_velocity = speeds_[2];
        else if (error < 250) new_velocity = speeds_[1];

        if (raw_diff < 0) new_velocity = -new_velocity;

        if (error < stop_threshold) break;

        write4B(ADDR_GOAL_VELOCITY, static_cast<uint32_t>(new_velocity), "Goal Velocity");
        std::cout << "Moving... Current: " << present_position
                  << ", Target: " << goal_position
                  << ", Error: " << error
                  << ", Velocity: " << new_velocity << "\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // Immediate stop (no ramp)
    write4B(ADDR_GOAL_VELOCITY, 0u, "Goal Velocity");
    std::cout << "Now at " << get_present_position() << ", stopping.\n";
}

double DxlServo::convert_servo_position_to_degrees(int servo_position) const {
    const double servo_values_per_degree = double(MAX_POSITION) / 360.0;
    double result = (servo_position - offset_center_) / (servo_values_per_degree * multiplier_);
    if (result > 180.0) result -= 360.0;
    return std::round(result * 100.0) / 100.0; // round to 2 decimals
}

void DxlServo::set_position(double pos_deg, int velocity) {
    if (operating_mode_ != OPERATING_MODE_POSITION) {
        throw std::runtime_error("Servo is not in position control mode.");
    }

    if (velocity >= 0) {
        write4B(ADDR_PROFILE_VELOCITY, static_cast<uint32_t>(velocity), "Profile Velocity");
    }
    write4B(ADDR_PROFILE_ACCELERATION, 50u, "Profile Acceleration");

    int goal_position = convert_degrees_to_servo_position(pos_deg);
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler_->write4ByteTxRx(
        portHandler_, DXL_ID, ADDR_GOAL_POSITION, static_cast<uint32_t>(goal_position), &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        throw std::runtime_error(std::string("Failed to set goal position: ")
                                 + packetHandler_->getTxRxResult(dxl_comm_result));
    }
    if (dxl_error) {
        throw std::runtime_error(std::string("Error from servo: ")
                                 + packetHandler_->getRxPacketError(dxl_error));
    }
}

void DxlServo::close() {
    if (portHandler_) {
        portHandler_->closePort();
        // std::cout << "Port closed.\n";
    }
}

int DxlServo::convert_degrees_to_servo_position(double pos_deg) const {
    const double servo_values_per_degree = double(MAX_POSITION) / 360.0;
    int result = int(offset_center_ + (pos_deg * servo_values_per_degree) * multiplier_);
    if (result < 0) result += MAX_POSITION;
    return result;
}

// IO helpers
void DxlServo::write2B(int addr, uint16_t val, const char* what) {
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler_->write2ByteTxRx(
        portHandler_, DXL_ID, addr, val, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        throw std::runtime_error(std::string(what) + " Error: "
                                 + packetHandler_->getTxRxResult(dxl_comm_result));
    }
    if (dxl_error) {
        throw std::runtime_error(std::string(what) + " Error: "
                                 + packetHandler_->getRxPacketError(dxl_error));
    }
}

void DxlServo::write4B(int addr, uint32_t val, const char* what) {
    uint8_t dxl_error = 0;
    int dxl_comm_result = packetHandler_->write4ByteTxRx(
        portHandler_, DXL_ID, addr, val, &dxl_error);
    if (dxl_comm_result != COMM_SUCCESS) {
        throw std::runtime_error(std::string(what) + " Error: "
                                 + packetHandler_->getTxRxResult(dxl_comm_result));
    }
    if (dxl_error) {
        throw std::runtime_error(std::string(what) + " Error: "
                                 + packetHandler_->getRxPacketError(dxl_error));
    }
}

